lib/controllers/auth-controller.js

```diff
@@ -0,0 +1,162 @@
+const Joi = require("joi");
+const { BAD_REQUEST, INTERNAL_SERVER_ERROR, OK, UNAUTHORIZED } = require("../utils/status-codes");
+const { isTemporaryEmail, hashOrVerifyPassword, generateToken, getMailOptions, getTransport } = require("../utils/auth-utils");
+const { userModel } = require("../models/user-model");
+
+exports.signup = async (req, res) => {
+    const passwordPattern = /^(?=.*[a-zA-Z])(?=.*[0-9])(?=.*[!@#$%^&*])[a-zA-Z0-9!@#$%^&*]{8,}$/;
+
+    const schema = Joi.object({
+        firstName: Joi.string().required(),
+        lastName: Joi.string().required(),
+        email: Joi.string().email().required(),
+        password: Joi.string().min(8).pattern(passwordPattern).required().messages({'string.pattern.base': 'Password must contain at least one number, one uppercase, one lowercase and one special character.'}),
+        confirmPassword: Joi.string().min(8).required(),
+        terms: Joi.boolean().valid(true).required().messages({'any.only': 'Terms and Conditions must be accepted'})
+    });
+
+    const { error, value } = schema.validate(req.body, {abortEarly: false})
+
+    // If error in form validation
+    if(error) {
+        let errorMessage = [];
+        error.details.forEach(detail => {
+            errorMessage.push(detail.message);
+        })
+        return res.status(BAD_REQUEST).send({ message: errorMessage});
+    }
+
+    // Checking if the email is  from a valid provider
+    if(isTemporaryEmail(value.email)) {
+        return res.status(BAD_REQUEST).send({message: "Invalid email provider"})
+    }
+
+    // Checking that the passwords match
+    if(value.password !== value.confirmPassword) {
+        return res.status(BAD_REQUEST).send({ message: 'Passwords do not match'})
+    }
+
+    const hashedPassword = await hashOrVerifyPassword(value.password);
+
+    let user = await userModel.findById({_id: value.email});
+
+    // Checking if user already exists
+    if(user) {
+        return res.status(BAD_REQUEST).send({ message: 'Email address is already in use'});
+    }
+
+    try {
+        // Create new user and save it to database
+        user = new userModel({
+            _id: value.email,
+            firstName: value.firstName,
+            lastName: value.lastName,
+            password: hashedPassword
+        });
+
+        await user.save();
+
+    } catch (error) {
+        console.error(error);
+        return res.status(INTERNAL_SERVER_ERROR).send({message: "Error while saving user"})
+    }
+
+    const accessToken = generateToken(value.email, user, true);
+
```

Comment:
@ElegantGracie how would the FE get access to the access tokens if you send via the headers? Do they have to the res.headers?

Reply:
I have removed all response headers and left only the response body in my new commit.